File: ts/src/value.ts
/**
 * Typed values and coercion from raw environment strings.
 *
 * Every variable in the environment arrives as a string. The schema declares
 * what that string is *supposed* to be; `coerce` turns it into a typed
 * {@link Value} or reports, in human-readable prose, why it could not.
 */

/** The declared type of a variable in the schema. */
export type VarType =
  | "string"
  | "int"
  | "float"
  | "bool"
  | "url"
  | "port"
  | "enum"
  | "duration";

/** The default type when `type` is omitted in the schema. */
export const DEFAULT_VAR_TYPE: VarType = "string";

/** All recognised variable types, for schema validation and diagnostics. */
export const VAR_TYPES: readonly VarType[] = [
  "string",
  "int",
  "float",
  "bool",
  "url",
  "port",
  "enum",
  "duration",
];

/**
 * A successfully coerced value. The discriminated union mirrors the Rust
 * `Value` enum: each variant carries the native representation of its type.
 *
 * `duration` values carry their magnitude in **milliseconds** (matching the
 * Rust port's `Duration` semantics, which renders as `<n>ms`).
 */
export type Value =
  | { readonly kind: "string"; readonly value: string }
  | { readonly kind: "int"; readonly value: number }
  | { readonly kind: "float"; readonly value: number }
  | { readonly kind: "bool"; readonly value: boolean }
  | { readonly kind: "url"; readonly value: string }
  | { readonly kind: "port"; readonly value: number }
  | { readonly kind: "duration"; readonly millis: number };

/**
 * Numeric magnitude for `min`/`max` checks, if the value is comparable.
 * Durations compare by their value in **seconds** (matching the Rust port,
 * which uses `Duration::as_secs_f64`).
 */
export function asNumber(value: Value): number | undefined {
  switch (value.kind) {
    case "int":
    case "float":
    case "port":
      return value.value;
    case "duration":
      return value.millis / 1000;
    default:
      return undefined;
  }
}

/** Render a value as the string a human (or a JSON report) should see. */
export function valueToString(value: Value): string {
  switch (value.kind) {
    case "string":
    case "url":
      return value.value;
    case "int":
    case "port":
      return String(value.value);
    case "float":
      return formatFloat(value.value);
    case "bool":
      return value.value ? "true" : "false";
    case "duration":
      return `${Math.round(value.millis)}ms`;
  }
}

/**
 * Format a float the way Rust's `{}` does: integral values keep no decimal
 * point unless they were genuinely fractional (e.g. `3` not `3.0`, `3.5` as is).
 */
function formatFloat(n: number): string {
  if (Number.isInteger(n)) return String(n);
  return String(n);
}

/**
 * Coerce a raw string into the requested {@link VarType}.
 *
 * On success returns `{ ok: true, value }`; on failure `{ ok: false, error }`
 * with a human-readable message describing the expectation.
 */
export type CoerceResult =
  | { readonly ok: true; readonly value: Value }
  | { readonly ok: false; readonly error: string };

export function coerce(raw: string, ty: VarType): CoerceResult {
  switch (ty) {
    case "string":
      return { ok: true, value: { kind: "string", value: raw } };
    case "int": {
      const n = parseInteger(raw.trim());
      return n === undefined
        ? { ok: false, error: `expected an integer, got ${quote(raw)}` }
        : { ok: true, value: { kind: "int", value: n } };
    }
    case "float": {
      const n = parseFloatStrict(raw.trim());
      return n === undefined
        ? { ok: false, error: `expected a float, got ${quote(raw)}` }
        : { ok: true, value: { kind: "float", value: n } };
    }
    case "bool": {
      const b = parseBool(raw);
      return b === undefined
        ? {
            ok: false,
            error: `expected a boolean (true/false/1/0/yes/no/on/off), got ${quote(raw)}`,
          }
        : { ok: true, value: { kind: "bool", value: b } };
    }
    case "port": {
      const p = parsePort(raw);
      return p === undefined
        ? { ok: false, error: `expected a port in 1..=65535, got ${quote(raw)}` }
        : { ok: true, value: { kind: "port", value: p } };
    }
    case "url":
      return isUrl(raw)
        ? { ok: true, value: { kind: "url", value: raw } }
        : { ok: false, error: `expected a URL with scheme://authority, got ${quote(raw)}` };
    case "enum":
      // Enum values are validated against `values` by the schema layer; here we
      // simply carry the string through.
      return { ok: true, value: { kind: "string", value: raw } };
    case "duration": {
      const ms = parseDuration(raw);
      return ms === undefined
        ? { ok: false, error: `expected a duration like 30s/5m/2h, got ${quote(raw)}` }
        : { ok: true, value: { kind: "duration", millis: ms } };
    }
  }
}

/** Quote a string the way Rust's `{:?}` does, for parity in error messages. */
function quote(s: string): string {
  return JSON.stringify(s);
}

/** Parse a signed 64-bit-range integer. Rejects floats, NaN, and junk. */
function parseInteger(s: string): number | undefined {
  if (!/^[+-]?\d+$/.test(s)) return undefined;
  const n = Number(s);
  return Number.isSafeInteger(n) ? n : undefined;
}

/** Parse a finite float. Rejects empty strings, NaN, and trailing junk. */
function parseFloatStrict(s: string): number | undefined {
  if (s.length === 0) return undefined;
  // Reject Rust-incompatible spellings while accepting standard float syntax.
  if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(s) && !/^[+-]?(inf|infinity)$/i.test(s)) {
    return undefined;
  }
  const n = Number(s);
  return Number.isNaN(n) ? undefined : n;
}

function parseBool(raw: string): boolean | undefined {
  switch (raw.trim().toLowerCase()) {
    case "true":
    case "1":
    case "yes":
    case "on":
      return true;
    case "false":
    case "0":
    case "no":
    case "off":
      return false;
    default:
      return undefined;
  }
}

function parsePort(raw: string): number | undefined {
  const s = raw.trim();
  if (!/^\d+$/.test(s)) return undefined;
  const n = Number(s);
  return n >= 1 && n <= 65535 ? n : undefined;
}

/** Minimal, dependency-free URL shape check: `scheme://authority[...]`. */
export function isUrl(raw: string): boolean {
  const s = raw.trim();
  const sep = s.indexOf("://");
  if (sep < 0) return false;
  const scheme = s.slice(0, sep);
  const rest = s.slice(sep + 3);
  if (scheme.length === 0) return false;
  if (!/^[A-Za-z][A-Za-z0-9+.-]*$/.test(scheme)) return false;
  // Authority must be non-empty and not start with a path/query separator.
  const authority = rest.split(/[/?#]/, 1)[0] ?? "";
  return authority.length > 0;
}

/**
 * Parse a human duration into milliseconds. Supported suffixes: `ms`, `s`,
 * `m`, `h`, `d`. A bare number is interpreted as seconds.
 */
export function parseDuration(raw: string): number | undefined {
  const s = raw.trim();
  if (s.length === 0) return undefined;

  let num: string;
  let multMs: number;
  if (s.endsWith("ms")) {
    num = s.slice(0, -2);
    multMs = 1;
  } else if (s.endsWith("s")) {
    num = s.slice(0, -1);
    multMs = 1_000;
  } else if (s.endsWith("m")) {
    num = s.slice(0, -1);
    multMs = 60_000;
  } else if (s.endsWith("h")) {
    num = s.slice(0, -1);
    multMs = 3_600_000;
  } else if (s.endsWith("d")) {
    num = s.slice(0, -1);
    multMs = 86_400_000;
  } else {
    num = s;
    multMs = 1_000;
  }

  const value = parseFloatStrict(num.trim());
  if (value === undefined || value < 0 || !Number.isFinite(value)) return undefined;
  return value * multMs;
}

File: rust/src/value.rs
//! Typed values and coercion from raw environment strings.

use std::fmt;
use std::time::Duration;

use serde::Deserialize;

/// The declared type of a variable in the schema.
#[derive(Debug, Deserialize, Clone, Copy, PartialEq, Eq, Default)]
#[serde(rename_all = "lowercase")]
pub enum VarType {
    /// Any UTF-8 string (the default when `type` is omitted).
    #[default]
    String,
    /// A signed 64-bit integer.
    Int,
    /// A 64-bit float.
    Float,
    /// A boolean: `true/false`, `1/0`, `yes/no`, `on/off` (case-insensitive).
    Bool,
    /// A URL with a scheme and authority, e.g. `https://host[:port]/path`.
    Url,
    /// A TCP/UDP port in the range `1..=65535`.
    Port,
    /// One of an explicit set of allowed string values (`values = [...]`).
    Enum,
    /// A human duration such as `500ms`, `30s`, `5m`, `2h`, `1d`.
    Duration,
}

impl fmt::Display for VarType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            VarType::String => "string",
            VarType::Int => "int",
            VarType::Float => "float",
            VarType::Bool => "bool",
            VarType::Url => "url",
            VarType::Port => "port",
            VarType::Enum => "enum",
            VarType::Duration => "duration",
        };
        f.write_str(s)
    }
}

/// A successfully coerced value.
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Str(String),
    Int(i64),
    Float(f64),
    Bool(bool),
    Url(String),
    Port(u16),
    Duration(Duration),
}

impl Value {
    /// Numeric magnitude for `min`/`max` checks, if the value is comparable.
    pub(crate) fn as_number(&self) -> Option<f64> {
        match self {
            Value::Int(i) => Some(*i as f64),
            Value::Float(f) => Some(*f),
            Value::Port(p) => Some(*p as f64),
            Value::Duration(d) => Some(d.as_secs_f64()),
            _ => None,
        }
    }
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Value::Str(s) | Value::Url(s) => f.write_str(s),
            Value::Int(i) => write!(f, "{i}"),
            Value::Float(x) => write!(f, "{x}"),
            Value::Bool(b) => write!(f, "{b}"),
            Value::Port(p) => write!(f, "{p}"),
            Value::Duration(d) => write!(f, "{}ms", d.as_millis()),
        }
    }
}

/// Coerce a raw string into the requested [`VarType`].
///
/// Returns a human-readable error describing the expectation on failure.
pub fn coerce(raw: &str, ty: VarType) -> Result<Value, String> {
    match ty {
        VarType::String => Ok(Value::Str(raw.to_string())),
        VarType::Int => raw
            .trim()
            .parse::<i64>()
            .map(Value::Int)
            .map_err(|_| format!("expected an integer, got {raw:?}")),
        VarType::Float => raw
            .trim()
            .parse::<f64>()
            .map(Value::Float)
            .map_err(|_| format!("expected a float, got {raw:?}")),
        VarType::Bool => parse_bool(raw).map(Value::Bool).ok_or_else(|| {
            format!("expected a boolean (true/false/1/0/yes/no/on/off), got {raw:?}")
        }),
        VarType::Port => parse_port(raw)
            .map(Value::Port)
            .ok_or_else(|| format!("expected a port in 1..=65535, got {raw:?}")),
        VarType::Url => {
            if is_url(raw) {
                Ok(Value::Url(raw.to_string()))
            } else {
                Err(format!(
                    "expected a URL with scheme://authority, got {raw:?}"
                ))
            }
        }
        // Enum values are validated against `values` by the schema layer; here we
        // simply carry the string through.
        VarType::Enum => Ok(Value::Str(raw.to_string())),
        VarType::Duration => parse_duration(raw)
            .map(Value::Duration)
            .ok_or_else(|| format!("expected a duration like 30s/5m/2h, got {raw:?}")),
    }
}

fn parse_bool(raw: &str) -> Option<bool> {
    match raw.trim().to_ascii_lowercase().as_str() {
        "true" | "1" | "yes" | "on" => Some(true),
        "false" | "0" | "no" | "off" => Some(false),
        _ => None,
    }
}

fn parse_port(raw: &str) -> Option<u16> {
    match raw.trim().parse::<u32>() {
        Ok(n) if (1..=65535).contains(&n) => Some(n as u16),
        _ => None,
    }
}

/// Minimal, dependency-free URL shape check: `scheme://authority[...]`.
fn is_url(raw: &str) -> bool {
    let raw = raw.trim();
    let Some((scheme, rest)) = raw.split_once("://") else {
        return false;
    };
    if scheme.is_empty()
        || !scheme
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '+' | '-' | '.'))
    {
        return false;
    }
    if !scheme
        .chars()
        .next()
        .is_some_and(|c| c.is_ascii_alphabetic())
    {
        return false;
    }
    // Authority must be non-empty and not start with a path/query separator.
    let authority = rest.split(['/', '?', '#']).next().unwrap_or("");
    !authority.is_empty()
}

/// Parse a human duration. Supported suffixes: `ms`, `s`, `m`, `h`, `d`.
/// A bare number is interpreted as seconds.
fn parse_duration(raw: &str) -> Option<Duration> {
    let raw = raw.trim();
    if raw.is_empty() {
        return None;
    }
    let (num, mult_ms): (&str, f64) = if let Some(n) = raw.strip_suffix("ms") {
        (n, 1.0)
    } else if let Some(n) = raw.strip_suffix('s') {
        (n, 1_000.0)
    } else if let Some(n) = raw.strip_suffix('m') {
        (n, 60_000.0)
    } else if let Some(n) = raw.strip_suffix('h') {
        (n, 3_600_000.0)
    } else if let Some(n) = raw.strip_suffix('d') {
        (n, 86_400_000.0)
    } else {
        (raw, 1_000.0)
    };
    let value: f64 = num.trim().parse().ok()?;
    if value < 0.0 || !value.is_finite() {
        return None;
    }
    Some(Duration::from_secs_f64(value * mult_ms / 1_000.0))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn coerces_scalars() {
        assert_eq!(coerce("42", VarType::Int).unwrap(), Value::Int(42));
        assert_eq!(coerce("3.5", VarType::Float).unwrap(), Value::Float(3.5));
        assert_eq!(coerce("YES", VarType::Bool).unwrap(), Value::Bool(true));
        assert_eq!(coerce("off", VarType::Bool).unwrap(), Value::Bool(false));
        assert_eq!(coerce("8080", VarType::Port).unwrap(), Value::Port(8080));
    }

    #[test]
    fn rejects_bad_scalars() {
        assert!(coerce("not-int", VarType::Int).is_err());
        assert!(coerce("0", VarType::Port).is_err());
        assert!(coerce("70000", VarType::Port).is_err());
        assert!(coerce("maybe", VarType::Bool).is_err());
    }

    #[test]
    fn url_shape() {
        assert!(is_url("https://example.com"));
        assert!(is_url("postgres://user:pass@db:5432/app"));
        assert!(!is_url("example.com"));
        assert!(!is_url("://nohost"));
        assert!(!is_url("1http://x"));
        assert!(!is_url("http:///path"));
    }

    #[test]
    fn durations() {
        assert_eq!(parse_duration("500ms"), Some(Duration::from_millis(500)));
        assert_eq!(parse_duration("30s"), Some(Duration::from_secs(30)));
        assert_eq!(parse_duration("5m"), Some(Duration::from_secs(300)));
        assert_eq!(parse_duration("2h"), Some(Duration::from_secs(7200)));
        assert_eq!(parse_duration("1d"), Some(Duration::from_secs(86_400)));
        assert_eq!(parse_duration("10"), Some(Duration::from_secs(10)));
        assert_eq!(parse_duration("-1s"), None);
        assert_eq!(parse_duration("abc"), None);
    }
}
